### ISSUE TITLE
Proxy-Authorization is forwarded by proxy to the end server

### ISSUE THREAD

**Issue by dyang-branch** [eclipse/jetty.project#3973]
Hi Team,

   I am using jetty client 9.3.9 version. Currently, I am experiencing an issue where my squid proxy is sending the proxy-authorization header to the end server URI. Is there a way to avoid it?. I have confirmed that the Jetty HttpRequest will drop the header automatically if I hard coded the header in the request by using httpRequest.header("proxy-authorization","Basic Autooken") method.

Thanks
Daniel

**Comment by joakime** [comment]
> I am using jetty client 9.3.9 version

That's a very old version of Jetty.
Please use an up to date version of Jetty.
Consider version "9.4.19.v20190610"

There's been important fixes since 9.3.9, especially around security.

**Comment by dyang-branch** [comment]
Thanks @joakime , I am still curious  why the *proxy-authorization* is removed by jetty HttpRequest automatically if I hard coded the header in the request. However, if the request go throughs a proxy  , the header is not dropped.

**Comment by sbordet** [comment]
@dyang-branch it's not clear what Jetty component you are using.
`Proxy-Authorization` is a hop header (https://tools.ietf.org/html/rfc7235#section-4.4) so it should be removed by the proxy (and thus should never arrive to the server).

**Comment by dyang-branch** [comment]
Hi @sbordet and @joakime,

Thanks for your reply and after playing with different jetty version. I am still seeing the proxy.
**I am not sure whether it is a security issue or not**.  I am using the latest jetty-9.4.20 version and I am still seeing the proxy-authorization header by running the following code

```
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.HttpProxy;
import org.eclipse.jetty.client.Origin;
import org.eclipse.jetty.client.ProxyConfiguration;
import org.eclipse.jetty.client.api.ContentResponse;
import org.eclipse.jetty.client.util.BasicAuthentication;
import org.eclipse.jetty.util.ssl.SslContextFactory;

import java.net.URI;

public class Test {

    public static void main(String[] args) {
        try{
            String proxyHost = "Add valid Proxy URI here";
            int proxyPort = 80;
            String proxy_auth_username = "username";
            String proxy_auth_password = "password";
            SslContextFactory sslContextFactory = new SslContextFactory();
            HttpClient client = new HttpClient(sslContextFactory );
            ProxyConfiguration proxyConfig = client.getProxyConfiguration();
            HttpProxy proxy = new HttpProxy(new Origin.Address(proxyHost,proxyPort), false);
            client.getAuthenticationStore().addAuthentication(new BasicAuthentication(new URI("http://"+proxyHost+":"+proxyPort), "proxy", proxy_auth_username, proxy_auth_password));
            proxyConfig.getProxies().add(proxy);
            client.start();
            ContentResponse res = client.GET("http://postb.in/1566260858302-2209252382162");  //Get the latest postb.in link and check the header
            System.out.println(res.getStatus());
            System.out.println(res.getContentAsString());
            client.stop();}
        catch (Exception e){
            System.out.println("Error Message: "+ e.getMessage());
        }
    }
}
```
Please replace the link http://postb.in/1566260858302-2209252382162 with a latest link generated from https://postb.in and use a valid proxy with username and password configured.

Please let me know if you have problem to reproduce the issue.

**Comment by sbordet** [comment]
@dyang-branch it's not clear where you see the `Proxy-Authorization` header.
It's not clear if your proxy is Jetty based.
It's not clear if your server is Jetty based.
Please write a detailed diagram that shows where you see the header (in the client? in the proxy? in the server?).

**Comment by dyang-branch** [comment]
Hi @sbordet here is a detailed diagram and sorry for the delayed reply.
Our concern is that the **Proxy-authorization** header is used to authenticate the Squid proxy, why the it is forwarded to the public web server after Squid proxy validate it.  Please let me know if you need more information. Thank you so much.

https://www.draw.io/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1#R3VjbUtswEP2aPMLYcWzCIwQoMO0M08xw6UtHsTe2QJGCLCcOX9%2BVLcW3kAKF0OEl9h6vrNXZ3SM5PW80y79JMk9%2BiAhYr%2B9Eec876fX7rjsM8KKRVYkcOE4JxJJGxqkCxvQJDGjdMhpB2nBUQjBF500wFJxDqBoYkVIsm25TwZqzzkkMHWAcEtZFb2ikkhId%2Bk6FnwONEzuza9c3I9bZAGlCIrGsQd5pzxtJIVR5N8tHwDR5lpdy3NkzT9eBSeDqJQOu789%2Bs%2BVt5s9%2FLfKM%2B8NLNdkblG9ZEJaZBZtg1coyABESYkwhVSJiwQk7rdBjKTIegZ7GQavy%2BS7EHEEXwXtQamWySzIlEErUjJmn5Zx6omfXZqBUZDKELQuyNUJkDGqLn7fOAJYuiBkoucJxEhhRdNGMg5gaitd%2BFc14Y5h%2BBetuh%2FULrkAiZYiOQS4orrGdhybLy4QqGM9JQcYSe28TowuQCvLtnHY5sANM3ZrGxc4t7WXVBq6t7aTWAoHzQax5HdbGj6gOCF1Jka8%2BnTDP8f8zxvwOYz%2FhMYNUFSqqfxLQS8wmjIY6UJhoCrEAQRZxY9g64IBhpMcTxIJY3xV872EfJ0LSJ2wYwdHvHEikx7XSgISqJtepkuIBRoIJiQgXXIvIlDLWggijMUczBN0dCOj0UNTmI%2FNgRqOoUKBNyW2m%2Fz0awmt1hL8hv8GG%2FA4%2BKr%2BHL1BvHh3pbVDTyEiaYp4buYCcqtva%2FZ3ma9831klu6CuMVYPLd5HrZ2S4xqi%2FgVCLvVitzQxXgnJVJXTQbzas57cSVYZvRtW31taLfKdZGZ7TelG5HXVeVCR9vex%2F2E%2B6G8qV7ekbmIxNQ3%2ByPrZZ%2Bnx9dPsd3tx9fZhMIHyw%2B0qhc7IudIkRurYs9vXYiyn%2BXKN0Rdo6E3JJZLSW2kt9DsLrRh1GobSHgKYSf1VF7Q9bfRNsqIjDnVZEd8vcJqmG1oikScGL2xJXjjHdVoqqzbv6s0pfC%2Bv1AvvWc%2B5uBNYLWgJ78EaBDVp1crhbfbWi1KiKsvunoogzNH2G8GOmv%2Bgwh9506ji6wito2%2FFprSpkppuNT9J5MQY%2FHPVXLIbI%2F3piYyJObWS40jK4ctKvqyKDloqsFaOuIgfvoyJoVl%2FsZXlV%2F3t4p38A

**Comment by dyang-branch** [comment]
Team, could you please check whether this is a valid issue as we think it could be a security flaw.

**Comment by dyang-branch** [comment]
@sbordet, this thread seems not updated anymore. Could you please confirm whether this is a valid issue and we are still observing this issue with latest jetty library.

**Comment by joakime** [comment]
@dyang-branch your example code does not reproduce the issue.
and your diagram shows a squid proxy.
It makes no indication of how that squid proxy is setup.
doesn't indicate what kind of proxy you are using.
no detail on the proxy protocol (and version) configured and in use.
we are also missing details like if encryption is in play, or not.
we are missing details on how your authentication and authorization is setup on squid.
we haven't seen the request / response exchange details on both sides of the proxy yet.
in short, this question is missing massive amounts of detail.

our own testing of HttpClient shows that it does not forward the Proxy-Authorization past a proxy.
so far, there's no reproduction case, and no real details on what you see.

**Comment by sbordet** [comment]
@dyang-branch from your diagram is Squid that is sending the `Proxy-Authorization` header to the public web server. Seems to me a Squid issue or misconfiguration, unrelated to Jetty.
The HTTP specification clearly states that `Proxy-Authorization` should be dropped by the first proxy: https://tools.ietf.org/html/rfc7235#section-4.4.
